### PR TITLE
client/v2: remove unsafe json-iterator/reflect2

### LINKED ISSUE #13434: Remove use of unsafe json-iterator / reflect2 libraries
The etcd client currently depends on json-iterator via the client/v2 package:

https://github.com/etcd-io/etcd/blob/ef1f71a9f65a06b5810527731101de43a1b286d7/client/v2/json.go#L60-L73

This is still relied on by portions of the v3 libraries:

https://github.com/etcd-io/etcd/blob/ef1f71a9f65a06b5810527731101de43a1b286d7/server/etcdserver/api/v2discovery/discovery.go#L34

The approach taken by json-iterator and its reflect library is not stable (it pins to internal details of the stdlib reflect package), and requires maintenance per go release, which does not appear to be sustainable (https://github.com/golang/go/issues/48238#issuecomment-917000973)

In addition to the maintenance aspect, it is hard to reason about the safety of the implementation (https://github.com/golang/go/issues/48238#issuecomment-916331048)

Kubernetes has transitioned away from json-iterator and is using sigs.k8s.io/json - a library based on the stdlib json decoder that added in case-sensitivity and int-preserving behavior.

### PR DESCRIPTION
**Pull request by chaochn47**
Fix #13434

In addition to the reasons listed in #13434, the libraries were introduced by https://github.com/etcd-io/etcd/pull/10667 and driven by kubernetes usage. 

PTAL @liggitt

/cc @hexfusion

### CODE REVIEW

File: api/go.mod [hunk 1]
```diff
@@ -11,13 +11,6 @@ require (
 	google.golang.org/grpc v1.41.0
 )
 
-require (
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply api/go.mod]: yeah, I was using go 1.16. Updated the latest go.mod after upgrading to 1.17

**Comment by chaochn47** [comment]
Previous CI failed on 'bom' ref. https://github.com/etcd-io/etcd/runs/4079040518?check_suite_focus=true

Latest force commit fixed it and verified locally. 

```bash
GOARCH=amd64 PASSES='bom dep' ./test.sh
Running with --race
Starting at: Tue Nov  9 19:15:51 UTC 2021

'bom' started at Tue Nov  9 19:15:51 UTC 2021
Checking bill of materials...
% cp go.sum go.sum.tmp
% cp go.mod go.mod.tmp
% (cd tools/mod && gobin -p -m --mod=readonly github.com/coreos/license-bill-of-materials)
% /home/chaochn/workplace/EKS-etcd/src/etcd/tools/mod/.gobincache/github.com/coreos/license-bill-of-materials/@v/v0.0.0-20190913234955-13baff47494e/license-bill-of-materials --override-file ./bill-of-materials.override.json go.etcd.io/etcd/api/v3/... go.etcd.io/etcd/pkg/v3/... go.etcd.io/etcd/raft/v3/... go.etcd.io/etcd/client/pkg/v3/... go.etcd.io/etcd/client/v2/... go.etcd.io/etcd/client/v3/... go.etcd.io/etcd/server/v3/... go.etcd.io/etcd/etcdutl/v3/... go.etcd.io/etcd/etcdctl/v3/... go.etcd.io/etcd/tests/v3/... go.etcd.io/etcd/v3/...
% cp go.sum.tmp go.sum
% cp go.mod.tmp go.mod
'bom' completed at Tue Nov  9 19:15:55 UTC 2021

'dep' started at Tue Nov  9 19:15:55 UTC 2021
% (cd api && go list -m)
% (cd api && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/api/v3{{end}}{{end}} -m all)
% (cd pkg && go list -m)
% (cd pkg && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/pkg/v3{{end}}{{end}} -m all)
% (cd raft && go list -m)
% (cd raft && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/raft/v3{{end}}{{end}} -m all)
% (cd client/pkg && go list -m)
% (cd client/pkg && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/client/pkg/v3{{end}}{{end}} -m all)
% (cd client/v2 && go list -m)
% (cd client/v2 && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/client/v2{{end}}{{end}} -m all)
% (cd client/v3 && go list -m)
% (cd client/v3 && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/client/v3{{end}}{{end}} -m all)
% (cd server && go list -m)
% (cd server && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/server/v3{{end}}{{end}} -m all)
% (cd etcdutl && go list -m)
% (cd etcdutl && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/etcdutl/v3{{end}}{{end}} -m all)
% (cd etcdctl && go list -m)
% (cd etcdctl && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/etcdctl/v3{{end}}{{end}} -m all)
% (cd tests && go list -m)
% (cd tests && go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/tests/v3{{end}}{{end}} -m all)
% go list -m
% go list -f {{if not .Indirect}}{{if .Version}}{{.Path}},{{.Version}},go.etcd.io/etcd/v3{{end}}{{end}} -m all
SUCCESS: dependencies are consistent across modules
'dep' completed at Tue Nov  9 19:16:06 UTC 2021
SUCCESS
```

**Review by hexfusion — APPROVED** [review]
:+1:

**Comment by hexfusion** [comment]
thanks @chaochn47 !

**Pull request merged**